REGRESSION(289693@main): [SOUP] Internal Server Error on reddit.com

mcatanzaro opened the pull request:
#### b3028af5d2a48b45e857347318d6f0c045783dfb
<pre>
REGRESSION(289693@main): [SOUP] Internal Server Error on reddit.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=289129">https://bugs.webkit.org/show_bug.cgi?id=289129</a>

Reviewed by Patrick Griffis.

When a cookie changes, we currently check its domain and look for all
CookieChangeObservers with a hostname that matches exactly. This doesn&apos;t
work because this is not how cookies work. For example, when visiting
www.reddit.com, we receive cookies with domain .reddit.com and then
decide not to add them to the WebCookieCache because .reddit.com is not
equal to www.reddit.com. To fix this, loop through all the domains with
registered observers and use the existing soup_cookie_domain_matches()
function to decide what is a match and what is not.

Also: notify the cache of all existing cookies when configuring a new
cookie jar in setCookieStorage.

Also: add preprocessor guards in a couple locations where they were
missing.

* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::notifyCookie):
(WebCore::NetworkStorageSession::notifyCookieAdded):
(WebCore::NetworkStorageSession::notifyCookieDeleted):
(WebCore::NetworkStorageSession::cookiesDidChange):
(WebCore::NetworkStorageSession::setCookieStorage):
(WebCore::NetworkStorageSession::deleteAllCookies):

Canonical link: <a href="https://commits.webkit.org/292428@main">https://commits.webkit.org/292428@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f57b148970001b2c496f88d4906bb9fcbff1359a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95911 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5469 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100970 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46420 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15820 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73131 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30352 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98914 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86632 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53457 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11575 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4380 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45755 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4482 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102999 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22978 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16751 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82169 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23229 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82655 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81531 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20471 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26118 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3568 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16315 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22941 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28096 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22600 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26080 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24341 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -80,6 +80,44 @@ NetworkStorageSession::~NetworkStorageSession()
+        return;
+
+    auto host = String::fromUTF8(soup_cookie_get_domain(newCookie));
+    for (auto& host : m_cookieChangeObservers.keys()) {
```
The same variable name `host` is used here. Can the loop variable be made `const ref` afterwards?

mcatanzaro replied:
I think it can be const, yes. I'll try changing that.

spenap commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -80,6 +80,44 @@ NetworkStorageSession::~NetworkStorageSession()
+        return;
+
+    auto host = String::fromUTF8(soup_cookie_get_domain(oldCookie));
+    for (auto& host : m_cookieChangeObservers.keys()) {
```
This is the same, so I wonder if this is an idiom I don't recognise...

mcatanzaro replied:
It's just an error; thanks for flagging it. The first variable isn't doing anything, since it is shadowed by the second one. I need to to remove the first one.

spenap commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -80,6 +80,44 @@ NetworkStorageSession::~NetworkStorageSession()
+                break;
+
+            for (auto& observer : *observers)
+                observer.cookiesDeleted(host, { Cookie(oldCookie) });
```
seeing the code is the same in both cases I wonder if it makes sense to refactor into a method that takes a lambda (which would invoke cookiesAdded or cookiesDeleted appropriately)

mcatanzaro replied:
Fair point. I'll see whether that makes things simpler or not.

mcatanzaro replied:
This is not simpler.

spenap replied:
Fair enough, thanks for trying!

mcatanzaro committed:
REGRESSION(289693@main): [SOUP] Internal Server Error on reddit.com
https://bugs.webkit.org/show_bug.cgi?id=289129

Reviewed by Patrick Griffis.

When a cookie changes, we currently check its domain and look for all
CookieChangeObservers with a hostname that matches exactly. This doesn't
work because this is not how cookies work. For example, when visiting
www.reddit.com, we receive cookies with domain .reddit.com and then
decide not to add them to the WebCookieCache because .reddit.com is not
equal to www.reddit.com. To fix this, loop through all the domains with
registered observers and use the existing soup_cookie_domain_matches()
function to decide what is a match and what is not.

Also: notify the cache of all existing cookies when configuring a new
cookie jar in setCookieStorage.

Also: add preprocessor guards in a couple locations where they were
missing.

* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::notifyCookie):
(WebCore::NetworkStorageSession::notifyCookieAdded):
(WebCore::NetworkStorageSession::notifyCookieDeleted):
(WebCore::NetworkStorageSession::cookiesDidChange):
(WebCore::NetworkStorageSession::setCookieStorage):
(WebCore::NetworkStorageSession::deleteAllCookies):

Canonical link: https://commits.webkit.org/292428@main

csaavedra commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -80,6 +80,40 @@ NetworkStorageSession::~NetworkStorageSession()
 }
 
+#if HAVE(COOKIE_CHANGE_LISTENER_API)
+void NetworkStorageSession::notifyCookie(SoupCookie* cookie, bool added)
```
So when added is `false` that means it's being deleted? I don't think this is semantically sound. Maybe this is just me being pedantic tho :)

mcatanzaro replied:
Right. I can try refactoring it if you have a suggestion. I don't normally like bool parameters, but I also didn't want to add an enum to the header file just for this. The goal was just to avoid code duplication between these two functions.
